### PR TITLE
Code for Chapters 9-10

### PR DESCRIPTION
**Pull request by delbaoliveira**
**Search and Pagination**
- Decouple components, simplify table component by moving search and pagination components out. 
- Move shared logic up, from `table.tsx` to `page.tsx`
- Remove additional `fetchInvoiceCountBySearchTerm`
- Add debouncing

**Fetching Data**
- Add error handling and type definitions

**Mutating Data**
- Update seed script to use `uuid` 
- Refactor Data Fetching Functions
- Add Server Actions for creating, updating, deleting, invoices
- Add Zod

...and a bunch of other changes.

### CODE REVIEW

File: dashboard/15-final/app/lib/actions.ts [hunk 1]
```diff
@@ -1,11 +1,65 @@
...
+  redirect('/dashboard/invoices');
+}
+
+export async function deleteInvoice(formData: FormData) {
+  const { id } = DeleteInvoice.parse({
+    id: formData.get('id'),
+  });
+
+  await sql`DELETE FROM invoices WHERE id = ${id}`;
+  revalidatePath('/dashboard/invoices');
```
Comment (manovotny) [inline dashboard/15-final/app/lib/actions.ts:64]: Why no `redirect` here vs the other mutations?

File: dashboard/15-final/app/lib/data.ts [hunk 2]
```diff
@@ -1,30 +1,26 @@
...
-
-export async function fetchRevenueDelayed(): Promise<Revenue[]> {
+export async function fetchRevenue() {
   try {
     // We artificially delay a reponse for demo purposes.
     // Don't do this in real life :)
-    console.log('Fetching revenue data...');
-    await new Promise((resolve) => setTimeout(resolve, 3000));
+    // console.log('Fetching revenue data...');
+    // await new Promise((resolve) => setTimeout(resolve, 3000));
```
Comment (manovotny) [inline dashboard/15-final/app/lib/data.ts:15]: Do we want to keep these?

File: dashboard/15-final/app/lib/data.ts [hunk 3]
```diff
@@ -1,30 +1,26 @@
...
 
-    const revenueData = await sql`SELECT * FROM revenue`;
-    console.log('Data fetch complete after 3 seconds.');
+    const data = await sql`SELECT * FROM revenue`;
 
-    return revenueData.rows as Revenue[];
+    // console.log('Data fetch complete after 3 seconds.');
```
Comment (manovotny) [inline dashboard/15-final/app/lib/data.ts:19]: Do we want to keep this?

File: dashboard/15-final/app/lib/actions.ts (hunk 1 above)
Comment (delbaoliveira) [reply dashboard/15-final/app/lib/actions.ts:64]: Delete is on the same page as the invoices table, so we don't have to redirect.

File: dashboard/15-final/app/lib/data.ts (hunk 2 above)
Comment (delbaoliveira) [reply dashboard/15-final/app/lib/data.ts:15]: I ask them to try it out to simulate a slow data fetch. Though we don't keep is as a comment, we could just ask them to add it in.

**Review by manovotny — APPROVED** [review]

**Pull request merged**